fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/CascadesRuleCall.java

```diff
@@ -134,14 +141,53 @@ public void yield(@Nonnull ExpressionRef<? extends RelationalExpression> express
             for (RelationalExpression member : groupExpressionRef.getMembers()) {
                 if (root.insertFrom(member, groupExpressionRef)) {
                     newExpressions.add(member);
-                    aliasResolver.addExpression(expressionReference, member);
+                    traversal.addExpression(expressionReference, member);
                 }
             }
         } else {
             throw new RecordCoreArgumentException("found a non-group reference in an expression used by the Cascades planner");
         }
     }
 
+    @Nonnull
+    public ExpressionRef<? extends RelationalExpression> ref(@Nonnull final RelationalExpression expression) {
```

Comment:
This method allows memoization of a new expression, by looking at the references directly owned by the new expression. If there already is another expression using those references, it can be memoized. This approach works really well for algorithms following the classic dynamic programming paradigm such as during join enumeration and incremental predicate normalization.